File: catkin_ws/src/drone_gazebo/nodes/ros_tools/model_state_to_pose_stamped.py
#! /usr/bin/env python3

import fix_python3_path
import rospy
import tf.transformations as tft
from gazebo_msgs.msg import ModelStates
from geometry_msgs.msg import PoseStamped


class ModelStateToPoseStamped:
    def __init__(self):
        rospy.init_node("model_state_to_pose_stamped")
        self.model_name = rospy.get_param("~model_name", "wamv")
        self.relative_entity_name = rospy.get_param("~relative_entity_name", "world")  # default to "world"
        self.publish_rate = rospy.get_param("~publish_rate", 20)

        self.pose_msg = PoseStamped()
        
        self.pose_publisher = rospy.Publisher(f"/gazebo/{self.model_name}/pose", PoseStamped, queue_size=10)
        self.model_sub = rospy.Subscriber("/gazebo/model_states", ModelStates, self.model_callback)
        self.pub_timer = rospy.Timer(rospy.Duration(1.0 / self.publish_rate), self.publish_pose)


    def model_callback(self, msg):
        try:
            model_index = msg.name.index(self.model_name)
            rospy.loginfo_throttle(1, f"Model name: {msg.name[model_index]}")

            # Find the relative entity's index
            if self.relative_entity_name and self.relative_entity_name != "world":
                if self.relative_entity_name in msg.name:
                    relative_entity_index = msg.name.index(self.relative_entity_name)
                else:
                    rospy.logerr_throttle(1, f"Relative entity {self.relative_entity_name} not found.")
                    return

                # Compute the relative pose
                model_pose = msg.pose[model_index]
                relative_entity_pose = msg.pose[relative_entity_index]
                relative_pose = self.compute_relative_pose(model_pose, relative_entity_pose)

                # Update the PoseStamped message
                self.pose_msg.header.stamp = rospy.Time.now()
                self.pose_msg.header.frame_id = self.relative_entity_name + "/base_link"
                self.pose_msg.pose = relative_pose
            else:
                model_pose = msg.pose[model_index]
                self.pose_msg.header.stamp = rospy.Time.now()
                self.pose_msg.header.frame_id = "map"
                self.pose_msg.pose = model_pose
        except ValueError:
            rospy.logerr_throttle(1, f"Model with name {self.model_name} not found in the model states.")

    def compute_relative_pose(self, model_pose, relative_entity_pose):
        # Calculate the relative transformation
        model_quaternion = [
            model_pose.orientation.x,
            model_pose.orientation.y,
            model_pose.orientation.z,
            model_pose.orientation.w,
        ]
        relative_entity_quaternion = [
            relative_entity_pose.orientation.x,
            relative_entity_pose.orientation.y,
            relative_entity_pose.orientation.z,
            relative_entity_pose.orientation.w,
        ]

        relative_rotation = tft.quaternion_multiply(
            tft.quaternion_inverse(relative_entity_quaternion), model_quaternion
        )
        relative_position = tft.quaternion_multiply(
            tft.quaternion_multiply(
                tft.quaternion_inverse(relative_entity_quaternion),
                [
                    model_pose.position.x - relative_entity_pose.position.x,
                    model_pose.position.y - relative_entity_pose.position.y,
                    model_pose.position.z - relative_entity_pose.position.z,
                    0,
                ],
            ),
            relative_entity_quaternion,
        )[:3]

        # Convert to Pose
        relative_pose = model_pose
        relative_pose.position.x, relative_pose.position.y, relative_pose.position.z = relative_position
        (
            relative_pose.orientation.x,
            relative_pose.orientation.y,
            relative_pose.orientation.z,
            relative_pose.orientation.w,
        ) = relative_rotation

        return relative_pose

    def publish_pose(self, event):
        rospy.loginfo_throttle(1, f"Publishing pose for {self.model_name}")
        self.pose_publisher.publish(self.pose_msg)


if __name__ == "__main__":
    ModelStateToPoseStamped()
    rospy.spin()

File: catkin_ws/src/drone_gazebo/nodes/ros_tools/model_state_to_pose_stamped.cpp
#include <gazebo_msgs/ModelStates.h>
#include <geometry_msgs/PoseStamped.h>
#include <ros/ros.h>

#include <string>

class ModelStateToPoseStamped
{
  public:
    ModelStateToPoseStamped() : ready(false)
    {
        ros::NodeHandle nh("");
        ros::NodeHandle nh_private("~");
        nh_private.param("model_name", model_name, std::string("wamv"));
        nh_private.param("pose_topic_suffix", pose_topic_suffix, std::string("/pose_stamped"));
        nh_private.param("publish_rate", publish_rate, 100);

        std::string pose_topic = "/gazebo";
        if (!model_name.empty() && model_name[0] == '/')
        {
            pose_topic += model_name;
        }
        else
        {
            pose_topic += "/" + model_name;
        }
        if (!pose_topic_suffix.empty() && pose_topic_suffix[0] == '/')
        {
            pose_topic += pose_topic_suffix;
        }
        else
        {
            pose_topic += "/" + pose_topic_suffix;
        }

        pose_publisher = nh.advertise<geometry_msgs::PoseStamped>(pose_topic, 10);
        model_sub = nh.subscribe("/gazebo/model_states", 10, &ModelStateToPoseStamped::modelCallback, this);

        pose_msg.header.frame_id = "map";  // or whatever the fixed frame is in your context

        timer = nh.createTimer(ros::Duration(1.0 / publish_rate), &ModelStateToPoseStamped::publishPose, this);
    }

    void modelCallback(const gazebo_msgs::ModelStates::ConstPtr& msg)
    {
        try
        {
            auto model_index = std::find(msg->name.begin(), msg->name.end(), model_name) - msg->name.begin();
            if (model_index >= msg->name.size())
            {
                ROS_ERROR_STREAM_THROTTLE(1, "Model name " << model_name << " not found in the model states.");
                return;
            }

            ROS_INFO_STREAM_THROTTLE(1, "Model name: " << msg->name[model_index]);
            pose_msg.header.stamp = ros::Time::now();
            pose_msg.pose = msg->pose[model_index];
            ready = true;
        }
        catch (const std::exception& e)
        {
            ROS_ERROR_STREAM_THROTTLE(1, "Error: " << e.what());
        }
    }

    void publishPose(const ros::TimerEvent&)
    {
        if (!ready)
        {
            return;
        }
        ROS_INFO_STREAM_THROTTLE(1, "Publishing pose for " << model_name);
        pose_publisher.publish(pose_msg);
    }

  private:
    ros::Publisher pose_publisher;
    ros::Subscriber model_sub;
    ros::Timer timer;
    geometry_msgs::PoseStamped pose_msg;
    std::string model_name;
    std::string pose_topic_suffix;
    int publish_rate;
    bool ready;
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "model_state_to_pose_stamped");
    ModelStateToPoseStamped mstp;
    ros::spin();
    return 0;
}
